Store fixtures in files instead of DB

phillipm opened the pull request:
Fixtures used to be stored completely in the DB. This limited them to 1mb in size, which has become unrealistic. 
This PR moves fixture storage to a hybrid db / file approach: using the normal db setup but adding a column (`DatabaseHelper.FILE_COL`) that points to the filepath where the fixture is stored in a serialized, usually encrypted, manner. Also adds a table column for the encryption key (`DatabaseHelper.AES_COL`), so that encrypted io is contained in the storage layer.

Includes a database migration to the new setup.

Cross-requested with https://github.com/dimagi/javarosa/pull/209 and https://github.com/dimagi/commcare/pull/221

__Release Notes__: remove 1mb lookup table size restriction

phillipm commented on unit-tests/src/org/commcare/android/database/SqlFileBackedStorageTests.java:
```diff
@@ -0,0 +1,245 @@
+
+        // TODO PLM: this should return a result. I think the ':' isn't escaped properly in the query
+        Vector<FormInstance> shouldHaveFormResults = userFixtureStorage.getRecordsForValues(new String[]{FormInstance.META_ID}, new String[]{"commtrack:programs"});
+        Assert.assertTrue(shouldHaveFormResults.size() == 0);
```
Any thoughts on this ^?
A fixture with an instance id of  `commtrack:programs` should exist. I think the `:` is messing up the query. We should really look into this!

phillipm commented on unit-tests/src/org/commcare/android/database/SqlFileBackedStorageTests.java:
```diff
@@ -0,0 +1,272 @@
+                // seems to be required; otherwise iterator loops forever. Not
+                // sure if it is a robolectric bug or a bug in our iterator
+                // that comes up when we iterate and delete at the same time
+                break;
```
Worth noting ^

phillipm commented on app/src/org/commcare/android/database/SqlStorage.java:
```diff
@@ -413,12 +449,13 @@ public boolean isEmpty() {
+    protected Cursor getIterateCursor(SQLiteDatabase db, boolean includeData) {
+        String[] projection = includeData ? new String[]{DatabaseHelper.ID_COL, DatabaseHelper.DATA_COL} : new String[]{DatabaseHelper.ID_COL};
+        return db.query(table, projection, null, null, null, null, null);
+    }
```
Split this logic into a helper method so subclass `FileBackedSqlStorage` can override it with appropriate logic.

phillipm commented on app/src/org/commcare/android/database/SqlStorage.java:
```diff
@@ -199,43 +210,19 @@ public T getRecordForValues(String[] rawFieldNames, Object[] values) throws NoSu
-            if (c != null) {
-                c.close();
-            }
-        }
```
This was all duplicated in `getRecordForValues`

ctsims commented on app/src/org/commcare/android/crypt/EncryptionIO.java:
```diff
@@ -0,0 +1,95 @@
+            try {
+                Cipher cipher = Cipher.getInstance("AES");
+                cipher.init(Cipher.ENCRYPT_MODE, symetricKey);
+                return new CipherOutputStream(fos, cipher);
```
It seems odd to me that we buffer the read below, but not the write here. Is there a reason we don't?

phillipm replied:
I thought in the case where you have all the data already lying around you don't need to use buffered streams?
Since this is a util method used in several contexts, not just in when writing to a byte array to a file in `HybridFileBackedSqlStorage`, I'll add the buffered stream. But am I confused in saying that in the case of `HybridFileBackedSqlStorage` it isn't needed?

ctsims replied:
Writing to a buffered stream should still help with file writes. The data is in memory but we're still writing it in a way that is chunked out of convenience by whatever structure the incoming stream decides (IE: Regardless of whether it's 2 bytes or 134 bytes it'll still try to dump them whenever it has them). Using a buffered output steam lets the writer write in whatever way is convenient without actually starting the file I/O until it can write a nice clean chunk, which should speed things up on both ends.

Basically if one end of a stream is I/O backed that end of the stream should almost always be buffered unless we know for sure that it already is.

ctsims commented on app/src/org/commcare/android/crypt/EncryptionIO.java:
```diff
@@ -0,0 +1,95 @@
+            //files are smaller than their contents (padded encryption data, etc),
+            //so you can't actually know that's correct. We should be relying on the
+            //methods we use to read data to make sure it's all coming out.
+        } catch (FileNotFoundException e) {
```
You may have just copied this over, but is there a reason this returns null instead of forwarding the result like the previous method?

phillipm replied:
Changed to let FileNotFoundException propagate up.
Note the TODOs I've added https://github.com/dimagi/commcare-odk/pull/906/files#diff-475fa51f6592fd0eb3c922e364db1d6eR137 and https://github.com/dimagi/commcare-odk/pull/906/files#diff-09bcd67419bcbc8e281fe0458496ae74R47

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    private static File getUniqueFilename(File dbDir) {
+        String uniqueSuffix = "_" + (Math.random() * 5);
```
Can we just use GUID's for these? seems more reliable/consistent than making up a structure.

phillipm replied:
Like `PropertyUtils.genGUID`? Sounds good

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+                        Arrays.toString(values), Arrays.toString(rawFieldNames));
+            }
+            cur.moveToFirst();
+            byte[] serializedObj = cur.getBlob(cur.getColumnIndexOrThrow(DatabaseHelper.DATA_COL));
```
This logic seems replicated in quite a few places (read object, check null, deserialize from either file or blob),  like readBytes below and getRecordsForValues above

phillipm replied:
The logic is slightly different in each case:
 - in `getRecordsForValues` I would guess we want to compute the column indexes outside of the `while` loop
 - in `getRecordForValues` we avoid loading the entire byte array by using an input stream
 - while in `readBytes` we directly read out of the file into a byte array.

They could all be unified but I suspect it would lead to slight performance losses. Your call since you understand these implications better than myself.

ctsims replied:
Ah, ok,  I missed those subtleties, that sounds like a good justification to me.

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+
+            db.setTransactionSuccessful();
+        } catch (IOException e) {
+            e.printStackTrace();
```
? This should absolutely throw this exception, shouldn't it?

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+                    byte[] key = generateKeyAndAdd(contentValues);
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.FILE_COL, contentValues);
+
+                    writeStreamToFile(bos, dataFile.getAbsolutePath(), key);
```
There's no way to undo this op, currently. An atomic structure for this method would be

* Create DB row/record for "Remove file at PATH on next clean"
* Open Transaction
* Write Row
* Delete "Remove file" row
* Close transaction

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+            bos.writeTo(fileOutputStream);
+        } finally {
+            if (fileOutputStream != null) {
+                fileOutputStream.close();
```
My guess is we should wrap this in an IOException wrapper, since we generally swallow stream close IOExceptions. (The method signature is eating it right now)

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+            updatedContentValues.put(DatabaseHelper.AES_COL, (byte[])null);
+
+            File dataFile = new File(filename);
+            dataFile.delete();
```
If this occurs but the transaction fails our record is now invalid. I would add this to a queue somewhere that manages files to delete so we can't end up in a state where the record becomes corrupted

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+        db.update(table, updatedContentValues,
+                DatabaseHelper.ID_COL + "=?", new String[]{String.valueOf(id)});
+
+        writeStreamToFile(bos, filename, fileEncryptionKey);
```
Same comments as above re:atomicity

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+        try {
+            String filename = HybridFileBackedSqlHelpers.getEntryFilename(helper, table, id);
+            if (filename != null) {
+                File dataFile = new File(filename);
```
Same comments as above re:atomicity

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,544 @@
+            SQLiteDatabase db = getDbOrThrow();
+            db.beginTransaction();
+            try {
+                HybridFileBackedSqlHelpers.removeFiles(ids, helper, table);
```
Same comments as above re:atomicity

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedStorageIterator.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public T nextRecord() {
+        byte[] blob =
```
This doesn't seem to have a path to follow if the "includeData" flag was set to false.

phillipm replied:
Neither does the normal SqlIterator. I think the onus is currently on the caller to invoke this correctly.

ctsims replied:
Ah, ok. I thought the SqlIterator fetched on demand, if that's not true, all set on this.

ctsims commented on app/src/org/commcare/android/database/SqlStorage.java:
```diff
@@ -45,26 +47,35 @@
-        } catch (IllegalAccessException | InstantiationException ie) {
+            e = ctype.newInstance();
+        } catch (IllegalAccessException ie) {
+            ie.printStackTrace();
```
Odd, why did you unfold these?

phillipm replied:
"Multi-catch with these reflection exceptions requires API level 19"

phillipm replied:
I mistakenly folded them in a previous PR...

ctsims replied:
Oh right, obnoxious. Cool, thanks!

ctsims commented on app/src/org/commcare/android/database/migration/FixtureSerializationMigration.java:
```diff
@@ -48,16 +55,27 @@ public static boolean migrateFixtureDbBytes(SQLiteDatabase db, Context c) {
+            int count = 0;
+            for (SqlStorageIterator i = oldUserFixtureStorage.iterate(false); i.hasMore(); ) {
                 int id = i.nextID();
+                Log.d(TAG, "migrating fixture " + count++);
```
Seems dangerously cute. I'd pull it out, lest someone rely on count in the future, and someone else removes the log statement. Log statements don't feel like they should have side effects.

ctsims commented on app/src/org/commcare/android/database/user/UserDatabaseUpgrader.java:
```diff
@@ -214,7 +217,33 @@ private boolean upgradeSevenEight(SQLiteDatabase db, int oldVersion, int newVers
      */
     private boolean upgradeEightNine(SQLiteDatabase db, int oldVersion, int newVersion) {
-        return FixtureSerializationMigration.migrateFixtureDbBytes(db, c);
+        Log.d(TAG, "starting user fixture migration");
```
Is this identical to the other method other than one or two parameters? Feels like there's a lot of repeated code.

phillipm replied:
What other method? `FixtureSerializationMigration.migrateFixtureDbBytes` is still called below.

ctsims replied:
Similar to AppDatabaseUpgrader.java's update method which seems to do the ~same thing with slightly different args, it looks like?

ctsims commented on app/src/org/commcare/android/db/legacy/LegacySqlIndexedStorageUtility.java:
```diff
@@ -126,16 +123,17 @@ public T getRecordForValues(String[] rawFieldNames, Object[] values) throws NoSu
+        int dbId = c.getInt(c.getColumnIndexOrThrow(DatabaseHelper.ID_COL));
         c.close();
-        return newObject(data);
+        return newObject(data, dbId);
```
I'd prefer we move the newObject code in its old form to this class rather than change how this class functions in general. This was a snapshot of a procedural generation/translation process so it's really sketch for us to make changes to that process over time, and cleaner for the code to represent what the legacy object used to do

ctsims commented on app/src/org/commcare/dalvik/application/CommCareApplication.java:
```diff
@@ -701,12 +702,36 @@ public SQLiteDatabase getHandle() {
+     */
+    public String getUserDbDir() {
+        try {
+            return currentApp.storageRoot() + currentApp + "/" + getSession().getLoggedInUser().getUniqueId() + "/";
```
This shouldn't actually be necessary, should it? Isn't there already a sandbox for user specific data or is that partition only in the database?

phillipm replied:
I didn't see anything along those lines anywhere. Do you know where I should look?

ctsims replied:
Hm, it must be that we store things in one big directory and then only load/decrypt the files that have relevant DB roots. Seems reasonable to partition if we can.

ctsims commented on unit-tests/src/org/commcare/dalvik/application/CommCareTestApp.java:
```diff
@@ -0,0 +1,24 @@
+import org.commcare.android.database.UnencryptedHybridFileBackedSqlStorageMock;
+import org.javarosa.core.services.storage.Persistable;
+
+/**
```
Class level comment would be welcome here

phillipm replied:
Left one that is sort of vague, would appreciate suggestions for ways to make it clearer.

ctsims replied:
that clarified it for me. Thanks!

ctsims commented on app/src/org/commcare/dalvik/application/CommCareApplication.java:
```diff
@@ -701,12 +702,36 @@ public SQLiteDatabase getHandle() {
+        try {
+            return currentApp.storageRoot() + currentApp + "/" + getSession().getLoggedInUser().getUniqueId() + "/";
+        } catch (SessionUnavailableException e) {
+            return currentApp.storageRoot() + currentApp + "/";
```
It worries me a lot that we get the DB in two different ways here depending on whether the session is available, since it feels like a process could try to get the storage and get a fully _different_ location when someone isn't logged in, then log in and the fixture they were parsing is gone. Seems a bit safer to error out here

phillipm commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,570 @@
+                    insertedId = db.insertOrThrow(table, DatabaseHelper.FILE_COL, contentValues);
+
+                    writeStreamToFile(bos, dataFilePath, key);
+                    unsetFileAsOrphan(db, dataFilePath);
```
Before I apply this change to other db ops: Does the ordering with `setFileAsOrphan`/`unsetFileAsOrphan` look correct in terms of making `write` atomic?

phillipm commented:
@ctsims Think I addressed your concerns with making db events atomic. I still need to find a good place and time to clear the orphaned file table, but otherwise should be good to review again.
Any suggestions on when/where to perform the orphan file clearing?

phillipm commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,573 @@
+            db.endTransaction();
+        }
+
+        FileUtil.deleteFileOrDir(dbDir);
```
With the remove db methods I elected to perform file removal after the transaction completed. I'm operating under the assumption that if the transaction isn't set to successful then `endTransaction` will raise a runtime exception and the file deletion logic will never be reached. Wanted to double check: this assumption is correct, right?

ctsims replied:
I think the bigger problem is if something (anything) kills the process after end transaction but before deleting the dir, which will result in the files never going away, since our record of them is gone. Wouldn't it make more sense to delete the full directory after deleting the records from the table, but before the transaction ends?

phillipm replied:
But what if the file/dir deletion partially completes and then fails? Then the db entry would persist because the transaction rolled back, but the files would be missing

ctsims replied:
It's better for the db entry to persist and to try and silently fail to delete non existent files than for the db entries to disappear and the files to remain, though. We can keep trying to delete blank files, but if we remove the index of the files we need to wipe, we won't even know that we need to wipe anything.

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java:
```diff
@@ -0,0 +1,169 @@
+
+    private static File getUniqueFilename(File dbDir) {
+        String timeAsString =
+                DateTime.now().toString(DateTimeFormat.forPattern("MM_dd_yyyy_HH_mm_ss"));
```
I don't think we need this if we just get a long enough GUID below. Is there a specific reason we want to use the time?

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java:
```diff
@@ -0,0 +1,169 @@
+            }
+        }
+
+        removeFiles(files);
```
We're basically assuming that the filenames are globally unique, but if not, this approach can result in weird behavior like

Table[A,B]

Delete all files: A succeeds, B fails, File System State: B, Table State[A,B]

App checks whether 'A' exists on filesystem before putting something there, sees that it doesn't, adds a new file to slot A

purge triggered

"New" A is removed.

Seems like a fine tradeoff, but it seems worth being explicit that this helper system only works with globally unique names

phillipm replied:
Ah, yeah, entirely possible; made note of it in a comment.

ctsims commented:
@phillipm super minor comments and then g2g

ctsims commented on app/src/org/commcare/android/database/HybridFileBackedSqlHelpers.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    private static File getUniqueFilename(File dbDir) {
+        String filename = PropertyUtils.genGUID(14);
```
Sorry to keep being a pain in the ass about this. 14 chars is only about 72 bits of entropy, we wanna shoot for 128 as a general standard. There's actually a "generateUUID()" method specifically in the same class that will generate UUID v4 id's that I try to use for anything like this just to eliminate any concern.

phillipm commented on app/src/org/commcare/android/database/HybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,574 @@
+                                      String directoryName) {
+        super(tableName, classType, dbHelper);
+
+        final String subPath = GlobalConstants.FILE_CC_DB + directoryName + "/_" + tableName;
```
Prefix `tableName` part of the directory with `_` to prevent the issue where if `tableName = ""` and we call `removeAll` it removes subdirectories associated with other tables

phillipm commented on app/src/org/commcare/dalvik/application/CommCareApplication.java:
```diff
@@ -704,12 +707,32 @@ public SQLiteDatabase getHandle() {
+        try {
+            return getSession().getUserKeyRecordUUID();
+        } catch (SessionUnavailableException e) {
+            throw new RuntimeException(e);
```
Using the user key record uuid as the subdirectory name for a specific user level file-backed db directory. Made sure this same uuid is also used when performing db migrations from the old db setup to this new one.

phillipm commented on app/src/org/commcare/dalvik/application/CommCareApplication.java:
```diff
@@ -783,10 +806,10 @@ public boolean matches(UserKeyRecord ukr) {
-
-        sharedPreferencesEditor.commit();
+        // manually clear file-backed fixture storage to ensure files are removed
+        CommCareApplication._().getFileBackedUserStorage("fixture", FormInstance.class).removeAll();
```
Making sure to remove files for file-backed databases when a "clear user data" is performed

phillipm replied:
If we wanted to do this in one swoop, instead of calling it on all tables that are file-backed, we could delete the `app/some_app_id/file_db/ukr_uuid_for_user` directory

phillipm commented on app/src/org/commcare/android/database/UnencryptedHybridFileBackedSqlStorage.java:
```diff
@@ -0,0 +1,48 @@
+    public UnencryptedHybridFileBackedSqlStorage(String table,
+                                                 Class<? extends T> ctype,
+                                                 AndroidDbHelper helper) {
+        super(table, ctype, helper, "app_level");
```
Store all unencrypted file-backed db files in `app/some_app_id/file_db/app_level/_tableName`

phillipm commented:
@ctsims thanks for catching the :file_folder: :bomb: :boom: :boom: :boom:
Should be :ok_hand: :eyeglasses:
